src/signal_manager.sh

```diff
@@ -0,0 +1,30 @@
+. "$KW_LIB_DIR/kwio.sh" --source-only
+
+# This functions prints the default message when interrupting kw
+function default_interrupt_handler() {
+  say $'\n'$SEPARATOR
+  say Exiting...
+}
+
+# This function adds a new signal handler to an arbitrary signal
+#
+# @command The handler to be called when a signal is catched
+# @signals The signals to attach to this handler
+function signal_manager() {
+  local command="$1"
+  shift
+  local -a signals="$@"
+
+  if [ -z "$command" -o -z "${signals[@]}" ]; then
+    return 22 #EINVAL
+  fi
+
+  trap "$command" "${signals[@]}" 2>/dev/null
```

Comment:
Let me check If I understood your strategy here. Basically, you want that specific files to add this `signal_manager()` that points to the correct function for handling it, something like this:

```diff
diff --git a/src/mk.sh b/src/mk.sh
index 43dc291..62bbbcf 100755
--- a/src/mk.sh
+++ b/src/mk.sh
@@ -19,6 +19,7 @@
 
 . "$KW_LIB_DIR/vm.sh" --source-only # It includes kw_config_loader.sh and kwlib.sh
 . "$KW_LIB_DIR/remote.sh" --source-only
+. "$KW_LIB_DIR/signal_manager.sh"
 
 # Hash containing user options
 declare -A options_values
@@ -576,6 +577,8 @@ function mk_build()
     exit 0
   fi
 
+  signal_manager mk_signal_handler SIGINT SIGTERM
+
   menu_config=${menu_config:-"nconfig"}
   arch=${arch:-"x86_64"}
 
@@ -810,3 +813,9 @@ function deploy_parser_options()
       ;;
   esac
 }
+
+function mk_signal_handler()
+{
+  echo "+++++++++++++++++++++++++++++++"
+  warning "====> Yoyoyo... yoyoyo... it's me! Do something really nice here!"
+}
```

Am I correct?
If so, I think the way you call this trap will not work; you need to eval an expression. In other words, you need something like this:

```bash
eval "trap "$command" "${signals[@]}" 2>/dev/null"
```

Maybe you can use cmd_manager to invoke this trap (it might work as well). Actually, if you use `cmd_manager` you can use the TEST mode for testing this function.

Reply:
Why is it that we need to eval? It seems to be working in my machine this way.

Reply:
When something behaves weirdly in bash, you can check the return value by looking at `$?`; I basically added `echo "$?"` right after your trap, and I catch `1`, which means that something was wrong. Next, I removed `2>/dev/null` and I could see this error:
```
line 24: trap: SIGINT SIGTERM: invalid signal specification  
```
After a few minutes, I realize that we were using the trap in the wrong way since the syntax says:
``
trap [COMMANDS] [SIGNALS ...]
``
Since you double-quote `"${signals[@]}"`, it became a single parameter; i.e., trap did not find the signal `SIGINT SIGTERM`. If you just remove the double-quote, it will work because each array's entry will be interpreted as a different parameter (`SIGINT` `SIGTERM`). Finally, if you use eval, it will expand everything for you and the rest you probably understand at this point :)

I suggested you to use cmd_manager because you can easy test this part.

Reply:
About the double quotes around the array see [this](https://www.gnu.org/software/bash/manual/bash.html#Arrays):

> Any element of an array may be referenced using ${name[subscript]}. The braces are required to avoid conflicts with the shell’s filename expansion operators. If the subscript is ‘@’ or ‘\*’, the word expands to all members of the array name. These subscripts differ only when the word appears within double quotes. If the word is double-quoted, ${name[*]} expands to a single word with the value of each array member separated by the first character of the IFS variable, and ${name[@]} expands each element of name to a separate word.

tl;dr: `"${signals[@]}"` expands to as many words as there are in the array `signals`, and `"${signals[*]}` expands to only one word. The output of these two snippets make it very clear:

```
$ array=(a b c); for i in "${array[@]}"; do echo element: $i; done
element: a
element: b
element: c
```
```
$ array=(a b c); for i in "${array[*]}"; do echo element: $i; done
element: a b c
```
Anyway, you are getting an error that I am not. I included a test for exactly this purpose (`add_two_signals_Test()`). What is the output of `./run_tests.sh test signal_manager` for you?

Reply:
Excellent analysis, make sense to me. Btw, I noticed that you added `local -a signals=( "$@" )`, could you elaborate a little bit?

Reply:
I think that is where the error was coming from! Suppose we receive the following arguments:
`'echo goodbye!' SIGINT SIGTERM`

After the `shift` command, `"$@"` will look like:
`SIGINT SIGTERM`

Thus, to create an array with all the signals, we use
`local -a signals=( "$@" )`

Which will expand to:

`local -a signals=( "SIGINT" "SIGTERM" )`

Which is exactly what we want. If we ommited the parenthesis, we would have

`local -a signals=$@`

which would expand to

`local -a signals=SIGINT SIGTERM`

Which would discard the second signal.

Reply:
Nice! I really want to see a blog post on the Flusp website about this :)